src/theory/quantifiers/sygus/sygus_reconstruct.h

```diff
@@ -0,0 +1,300 @@
+ *           Obs[T] += (k, toBuiltIn(s[z]), s[z])
+ *       else if no s' in Pool[T] s.t.
+ *             rewrite(toBuiltIn(s')) * sigma = rewrite(toBuiltIn(s))
+ *           Pool[T] += s[z]
+ *           for each (k, t, null) in Obs[T]
+ *             Obs' += matchNewObs(k, s[z])
+ *     // match phase
+ *     while Obs' != {}
+ *       Obs'' = {}
+ *       for each (k, t, null) in Obs' // s = null for all triples in Obs'
+ *         Obs[T] += (k, t, null)
+ *         for each s[z] in Pool[T]
+ *           Obs'' += matchNewObs(k, s[z])
+ *       Obs' = Obs''
+ *
+ *   g = Sol[k_0]
+ * }
+ *
+ * matchNewObs(k, s[z]) returns Obs'
+ * {
+ *   u = rewrite(toBuiltIn(s[z]))
+ *   if match(u, t) == {toBuiltin(z) -> t'}
+ *     if forall t' exists (k'', t'', s'') in Obs[T] s.t. |=_X t'' = t'
+ *       markSolved(k, s{z -> s''})
+ *     else
+ *       let k' be a new variable of type : typeOf(z)
+ *       CandSol[k] += s{z -> k'}
+ *       Obs'[typeOf(z)] += (k', t', null)
+ * }
+ *
+ * markSolved(k, s)
+ * {
+ *   if Sol[k] != null
+ *     return
+ *   instantiate free variables of s with arbitrary sygus datatype values
+ *   Sol[k] = s
+ *   CandSols[k] += s
+ *   Stack = {k}
+ *   while Stack != {}
+ *     k' = pop(Stack)
+ *     for all k'' in CandSols keys
+ *       for all s'[k'] in CandSols[k'']
+ *         s'{k' -> Sol[k']}
+ *         if s' does not contain free variables in Obs
+ *           instantiate free variables of s' with arbitrary sygus datatype
+ *             values
+ *           Sol[k''] = s'
+ *           push(Stack, k'')
+ * }
+ */
+class SygusReconstruct : public expr::NotifyMatch
+{
+ public:
+  /**
+   * Constructor.
+   *
+   * @param tds Database for sygus terms
+   * @param s Statistics managed for the synth engine
+   */
+  SygusReconstruct(TermDbSygus* tds, SygusStatistics& s);
+
+  /** reconstruct solution
+   *
+   * Return (if possible) a sygus encoding of a node that is equivalent to sol
+   * and whose syntax matches the grammar corresponding to sygus datatype stn.
+   *
+   * For example, given:
+   *   sol = (* 2 x)
+   *   stn = A sygus datatype encoding the grammar
+   *           Start -> (c_PLUS Start Start) | c_x | c_0 | c_1
+   * This method may return (c_PLUS c_x c_x) and set reconstructed to 1.
+   *
+   * @param sol The target term
+   * @param stn The sygus datatype type encoding the syntax restrictions
+   * @param reconstructed The flag to update, set to 1 if we successfully return
+   *                      a node, otherwise it is set to -1
+   * @param enumLimit A value to limit the effort spent by this class (roughly
+   *                  equal to the number of intermediate terms to try)
+   * @return The reconstructed sygus term
+   */
+  Node reconstructSolution(Node sol,
+                           TypeNode stn,
+                           int& reconstructed,
+                           int enumLimit);
+
+ private:
+  /** match obligation `k`'s builtin term with pattern `sz`
```

Comment:
```suggestion
  /** Match obligation `k`'s builtin term with pattern `sz`
```

Also this is called `ob` now in the signature?